be/src/exec/pipeline/scan/morsel.cpp

```diff
@@ -229,5 +236,304 @@ Status PhysicalSplitMorselQueue::_init_segment() {
+                                           const vectorized::ShortKeyOptionPtr& upper) const {
+    // It is validated that any of endpoint is infinite or the original short key range is a point.
+    if (lower->is_infinite() || upper->is_infinite() ||
+        _block_ranges_per_seek_range[_range_idx].first == _block_ranges_per_seek_range[_range_idx].second) {
+        return true;
+    }
+
+    Slice lower_key = lower->short_key;
+    // Empty short key of start ShortKeyOption means it is the first splitted key range,
+    // so use start original short key to compare.
+    if (lower_key.empty()) {
+        lower_key = *_block_ranges_per_seek_range[_range_idx].first;
+    }
+
+    Slice upper_key = upper->short_key;
+    // Empty short key of end ShortKeyOption means it is the last splitted key range,
+    // so use end original short key to compare.
+    if (upper_key.empty()) {
+        auto end_iter = _block_ranges_per_seek_range[_range_idx].second;
+        --end_iter;
+        upper_key = *end_iter;
+    }
+
+    return lower_key.compare(upper_key) != 0;
+}
+
+vectorized::ShortKeyOptionPtr LogicalSplitMorselQueue::_create_range_lower() const {
+    // If it is the first splitted key range, the start point is the original start key.
+    if (_next_lower_block_iter == _block_ranges_per_seek_range[_range_idx].first) {
+        if (_tablet_seek_ranges.empty()) {
+            return std::make_unique<vectorized::ShortKeyOption>();
+        } else {
+            return std::make_unique<vectorized::ShortKeyOption>(&_tablet_seek_ranges[_range_idx].lower(),
+                                                                _tablet_seek_ranges[_range_idx].inclusive_lower());
+        }
+    } else {
+        Slice short_key = *_next_lower_block_iter;
+        return std::make_unique<vectorized::ShortKeyOption>(_short_key_schema, short_key, true);
+    }
+}
+
+vectorized::ShortKeyOptionPtr LogicalSplitMorselQueue::_create_range_upper() const {
+    // If it is the last splitted key range, the end point is the original end key.
+    if (_next_lower_block_iter == _block_ranges_per_seek_range[_range_idx].second) {
+        if (_tablet_seek_ranges.empty()) {
+            return std::make_unique<vectorized::ShortKeyOption>();
+        } else {
+            return std::make_unique<vectorized::ShortKeyOption>(&_tablet_seek_ranges[_range_idx].upper(),
+                                                                _tablet_seek_ranges[_range_idx].inclusive_upper());
+        }
+    } else {
+        Slice short_key = *_next_lower_block_iter;
+        return std::make_unique<vectorized::ShortKeyOption>(_short_key_schema, short_key, false);
+    }
+}
+
+bool LogicalSplitMorselQueue::_cur_tablet_finished() const {
+    return _range_idx >= _block_ranges_per_seek_range.size();
+}
+
+Rowset* LogicalSplitMorselQueue::_find_largest_rowset(const std::vector<RowsetSharedPtr>& rowsets) {
+    Rowset* largest_rowset = nullptr;
```

Comment:
Could assign the first rowset to largest_rowset, then we needn't largest_rowset == nullptr check?

Reply:
Great.